cleveref-usedon.dtx

```diff
@@ -817,43 +941,85 @@
 %	Throw an error, if an unrecognised option was used for the
 %	optional argument to this macro.
 %\iffalse
-%%                    % Throw an error, if an unrecognised option was used
-%%                    % for the optional argument to this macro.
+%%            % Throw an error, if an unrecognised option was used
+%%            % for the optional argument to this macro.
 %\fi
 %    \begin{macrocode}
             \msg_new:nnn {cleveref-usedon} { OptionSpellingError }
-                {   \MessageBreak
-                    Spelling~error~\msg_line_context:
-                    \MessageBreak
-                    Did~you~mean~to~pass~option\MessageBreak
-                    'UsedOn'~to~cref~or~Cref?
-                }
+              {
+                \MessageBreak
+                Spelling~error~\msg_line_context:
+                \MessageBreak
+                Did~you~mean~to~pass~option\MessageBreak
+                'UsedOn'~to~cref~or~Cref?
+              }
             \msg_fatal:nn { cleveref-usedon } { OptionSpellingError }
         }
     }
-}%
+  }
 %    \end{macrocode}
 %\end{macro}
 %
 %\begin{macro}{\@@_cref}
 %	This is just a wrapper around \pkg{cleveref}'s |\cref|.
 %	Additionally the \cs{@@_Processor} gets called.
 %    \begin{macrocode}
-\NewDocumentCommand{\@@_cref}{ s o m }{%
-    \IfBooleanTF{#1}{ \@@_origcref*{#3} }{ \@@_origcref{#3} }%
-    \@@_Processor[#2]{#3}
-}%
-%    \end{macrocode}
-%\end{macro}
-%
-%\begin{macro}{\@@_Cref}
-%	This is just a wrapper around \pkg{cleveref}'s |\Cref|.
-%	Additionally the \cs{@@_Processor} gets called.
-%    \begin{macrocode}
-\NewDocumentCommand{\@@_Cref}{ s o m }{%
-    \IfBooleanTF{#1}{ \@@_origCref*{#3} }{ \@@_origCref{#3} }%
-    \@@_Processor[#2]{#3}
-}%
+\bool_new:N \l_@@_already_specified_bool
+\clist_new:N \l_@@_args_clist
+\cs_new:Nn \@@_define_from_orig:NN
+  {
+    \NewDocumentCommand #1 { s o m }
+      {
+        \IfValueTF { ##2 }
+          {
+            \bool_set_false:N \l_@@_already_specified_bool
+            \clist_set:Nn \l_@@_args_clist { ##2 }
+            \clist_put_right:No \l_@@_args_clist { \l_@@_default_tl }
+            \clist_map_inline:Nn \l_@@_args_clist
+              {
+                \str_case:xnT { \str_foldcase:n { ####1 } }
+                  {
+                    {usedon} {}
+                    {uo}     {}
+                    {usedby} {}
+                    {ub}     {}
+                    {usedbyandon} {}
+                    {ubao}        {}
+                  }
+                  {
+                    \clist_remove_all:Nn \l_@@_args_clist { ####1 }
+                    \bool_if:NF \l_@@_already_specified_bool
+                      {
+                        \@@_Processor[####1]{##3}
+                      }
+                    \bool_set_true:N \l_@@_already_specified_bool
+                  }
+              }
+            \clist_if_empty:NTF \l_@@_args_clist
+              {
+                \IfBooleanTF{##1}{ #2*{##3} }{ #2{##3} }
```

Comment:
The syntax was hard to decipher here at first due to LaTeX3 convention for parameter names.
I think in a future commit, I will include DocStrings to each function explaining what each argument does.

Reply:
I usually just write it as comments following the definition, for example, in my package `crefthe`, there is
```latex
\cs_new_protected:Nn \__crefthe_cref_general:NNNnnN
  % #1 = star or not
  % #2 = -
  % #3 = +
  % #4 = key-value
  % #5 = label
  % #6 = original command, such as \crefthe_cref_original:w or \crefthe_Cref_original:w
  {
    % ...
  }
```
Your idea seems more formal, but since I am not familiar with DocStrings, I shall leave this part of documentation to you :)